File: gateway/router.py
from __future__ import annotations

import asyncio
import logging
import re
from pathlib import Path

from ..config import Config
from ..memory.workspace import bootstrap, build_system_prompt
from ..permissions import PermissionPolicy, PermissionRequired
from ..agent import loop as agent_loop
from .session import SessionStore

log = logging.getLogger(__name__)

_MAX_CHUNK = 1000  # chars per block when streaming long responses


def _chunk_response(text: str, max_chars: int = _MAX_CHUNK) -> list[str]:
    if len(text) <= max_chars:
        return [text]
    paragraphs = re.split(r"\n{2,}", text)
    chunks: list[str] = []
    current = ""
    for para in paragraphs:
        candidate = (current + "\n\n" + para).strip() if current else para
        if len(candidate) > max_chars and current:
            chunks.append(current)
            current = para
        else:
            current = candidate
    if current:
        chunks.append(current)
    return chunks


class Router:
    def __init__(self, config: Config):
        self._config = config
        db_path = Path.home() / ".nanoclaw" / "gateway.db"
        self._store = SessionStore(db_path)
        self._active: dict[str, asyncio.Lock] = {}
        bootstrap(config.workspace)

    async def on_message(self, channel: str, peer_id: str, text: str, deliver) -> None:
        key = f"{channel}:{peer_id}"
        if key not in self._active:
            self._active[key] = asyncio.Lock()
        lock = self._active[key]

        if lock.locked():
            log.info("Dropping message from %s — turn already in progress", key)
            return

        async with lock:
            session = self._store.get_or_create(peer_id, channel)
            system_prompt = build_system_prompt(self._config.workspace)

            if not session.messages:
                session.messages.append({"role": "system", "content": system_prompt})

            session.messages.append({"role": "user", "content": text})

            policy = PermissionPolicy.from_mode(self._config.permission_mode)

            try:
                response = await agent_loop.run(
                    session.messages,
                    self._config,
                    session_id=session.id,
                    policy=policy,
                )
            except PermissionRequired as e:
                response = f"[Permission required for {e.tool} — switch to auto mode or approve explicitly]"
            except Exception as e:
                log.exception("Agent error for %s", key)
                response = f"[Error: {e}]"

            session.messages.append({"role": "assistant", "content": response})
            self._store.save(session)

            for chunk in _chunk_response(response):
                await deliver(peer_id, chunk)

File: src/gateway/router.ts
import { homedir } from "os";
import { join } from "path";
import type { Config } from "../config.ts";
import { bootstrap, buildSystemPrompt } from "../memory/workspace.ts";
import { PermissionPolicy, PermissionRequired } from "../permissions.ts";
import { run as agentRun } from "../agent/loop.ts";
import { SessionStore } from "./session.ts";

const MAX_CHUNK = 1000; // chars per block when streaming long responses

export type DeliverFn = (peerId: string, text: string) => Promise<void>;
export type OnMessageFn = (
  channel: string,
  peerId: string,
  text: string,
  deliver: DeliverFn,
) => Promise<void>;

// Per-peer mutex with single-slot pending buffer.
// If a turn is running: store the latest fn in the slot (newest wins).
// When the running turn finishes: drain the slot once, then release.
// This prevents pile-up while ensuring the most recent message is never lost.
const _locks = new Map<string, Promise<void>>();
const _pending = new Map<string, () => Promise<void>>();

function withLock(key: string, fn: () => Promise<void>): Promise<void> {
  const current = _locks.get(key);
  if (current !== undefined) {
    _pending.set(key, fn);
    return current;
  }

  let resolve!: () => void;
  const next = new Promise<void>((r) => {
    resolve = r;
  });
  _locks.set(key, next);

  const run = async (): Promise<void> => {
    try {
      await fn();
      const queued = _pending.get(key);
      if (queued) {
        _pending.delete(key);
        await queued();
      }
    } finally {
      _locks.delete(key);
      resolve();
    }
  };

  return run();
}

function chunkResponse(text: string, maxChars: number = MAX_CHUNK): string[] {
  if (text.length <= maxChars) return [text];

  const paragraphs = text.split(/\n{2,}/);
  const chunks: string[] = [];
  let current = "";

  for (const para of paragraphs) {
    const candidate = current ? `${current}\n\n${para}`.trim() : para;
    if (candidate.length > maxChars && current) {
      chunks.push(current);
      current = para;
    } else {
      current = candidate;
    }
  }

  if (current) chunks.push(current);
  return chunks;
}

export class Router {
  private readonly store: SessionStore;

  constructor(private readonly config: Config) {
    const dbPath = join(homedir(), ".nanoclaw", "gateway.db");
    this.store = new SessionStore(dbPath);
    bootstrap(config.workspace);
  }

  async onMessage(
    channel: string,
    peerId: string,
    text: string,
    deliver: DeliverFn,
  ): Promise<void> {
    const key = `${channel}:${peerId}`;

    await withLock(key, async () => {
      const session = this.store.getOrCreate(peerId, channel);
      const systemPrompt = buildSystemPrompt(this.config.workspace);

      if (session.messages.length === 0) {
        session.messages.push({ role: "system", content: systemPrompt });
      }

      session.messages.push({ role: "user", content: text });

      const policy = PermissionPolicy.fromMode(this.config.permissionMode);

      let response: string;
      try {
        response = await agentRun(session.messages, this.config, {
          sessionId: session.id,
          policy,
        });
      } catch (err) {
        if (err instanceof PermissionRequired) {
          response = `[Permission required for ${err.tool} — switch to auto mode or approve explicitly]`;
        } else {
          console.error(`Agent error for ${key}:`, err);
          response = `[Error: ${err instanceof Error ? err.message : String(err)}]`;
        }
      }

      session.messages.push({ role: "assistant", content: response });
      this.store.save(session);

      for (const chunk of chunkResponse(response)) {
        await deliver(peerId, chunk);
      }
    });
  }
}
